code/utils_model.py

```diff
@@ -1,479 +1,509 @@
+          f"learning_rate: {config.args.learning_rate}\n"
+          f"batch_size: {config.args.batch_size}\n"
+          f"weight_decay: {config.args.weight_decay}\n"
+          f"learning_rate_decay: {config.args.learning_rate_decay}\n"
+          f"resume_checkpoint: {config.args.resume_checkpoint}\n"
+          f"resume_checkpoint_path (only if resume_checkpoint is true): "
+          f"{config.resume_checkpoint_path}\n"
+          f"save_interval: {config.args.save_interval}\n"
+          f"output in checkpoints_folder: {config.args.checkpoints_folder}\n"
+          f"pretrain: {config.args.pretrain}\n"
+          f"log_csv: {config.log_csv}\n\n")
 
 
 ###########################################
 #          MAIN TRAIN FUNCTION            #
 ###########################################
 
 
-# helper function for training resnet
-def train_helper(model, dataloaders, device, dataset_sizes, criterion,
-                 optimizer, scheduler, num_epochs, save_interval, writer):
-
+def train_helper(model: torchvision.models.resnet.ResNet,
+                 dataloaders: Dict[str, torch.utils.data.DataLoader],
+                 dataset_sizes: Dict[str, int],
+                 criterion: torch.nn.modules.loss, optimizer: torch.optim,
+                 scheduler: torch.optim.lr_scheduler, num_epochs: int,
+                 writer: IO) -> None:
+    """
+    Function for training ResNet.
+
+    Args:
+        model: ResNet model for training.
+        dataloaders: Dataloaders for IO pipeline.
+        dataset_sizes: Sizes of the training and validation dataset.
+        criterion: Metric used for calculating loss.
+        optimizer: Optimizer to use for gradient descent.
+        scheduler: Scheduler to use for learning rate decay.
+        num_epochs: Number of epochs to train the network.
+        writer: Writer to write logging information.
+    """
     since = time.time()
 
-    # each epoch
-    for epoch in range(num_epochs):
-
-        # training phase
-        phase = 'train'
-        model.train()
+    # Initialize all the tensors to be used in training and validation.
+    # Do this outside the loop since it will be written over entirely at each
+    # epoch and doesn't need to be reallocated each time.
+    train_all_labels = torch.empty(size=(dataset_sizes["train"], ),
+                                   dtype=torch.long,
+                                   device=config.device)
+    train_all_predicts = torch.empty(size=(dataset_sizes["train"], ),
+                                     dtype=torch.long,
+                                     device=config.device)
+    val_all_labels = torch.empty(size=(dataset_sizes["val"], ),
+                                 dtype=torch.long,
+                                 device=config.device)
+    val_all_predicts = torch.empty(size=(dataset_sizes["val"], ),
+                                   dtype=torch.long,
+                                   device=config.device)
```

Comment:
@ntomita I didn't think about using CPU memory.  I didn't want to do a transfer back to CPU memory from the GPU.  This probably should be done on the CPU to avoid memory usage on the GPU.

Reply:
Tensors now stored in CPU memory.